### PR TITLE
Add widgets for setting histogram bins

### LINKED ISSUE #239: Allow adjustment of histogram bins
It would be great to allow:
- Adjustment of histogram bin width / locations via the UI
- Using integer bin limits for integer images (e.g. `uint8`, `uint16`)

For example, if you show the coins image in Napari (`File > Open Sample > napari builtins > Coins`) and view the histogram, it has an odd 'spiky' appearance where every other bin is higher or lower:
Screenshot 2023-12-14 121238

I think this is because the limits of the bins aren't set to integer values (even though this image is `uint8` and therefore only contains positive integers). E.g. I think the bins are `np.linspace(1, 252, 100)` according to [here](https://github.com/matplotlib/napari-matplotlib/blob/main/src/napari_matplotlib/histogram.py#L51), so for example bin 20 onwards is: `51.7, 54.2, 56.7, 59.3, 61.8`, which means the bins accept pixel values of (52, 53, 54), (55, 56), (57, 58, 59), (60, 61)... This alternating of 2 and 3 pixel values in each bin gives this spiky appearance which is a bit misleading to the actual values. Would it be possible to set integer bin edges, or allow this to be set as an option? Thanks!

### PR DESCRIPTION
**Pull request by p-j-smith**
Fixes #239

- added `QDoubleSpinBox` widgets for setting the start and stop edges of the histogram bins (default to the min and max data values, same as existing values used)
- added a `QSpinBox` widget for setting the number of bins to use (defaults to 100, same as the existing value used)
- add callbacks to re-draw figure when value of widgets change
- ~~set the `dtype` for `np.linspace` when creating bins (so integer bin limits are used for integer images). This caused the `test_histogram_2D` test to fail (as the test layer has integer data but float bins were being used) so I've updated the baseline image~~ - this is now done in #244 
- added a test for the `HistogramWidget` when the bin parameters are set from the widgets
- add `start`, `stop`, and `num_bins` parameters to `_get_bins` function so that bin parameters set in the widgets can be used for settings bins
- show notification if the bin parameters requested by the user have been modified before plotting (this can happen with integer data as we need to make sure the bins have integer widths)

https://github.com/matplotlib/napari-matplotlib/assets/29753790/876ab0e9-0b46-4efd-b408-c4ffa184d799

### CODE REVIEW

**Comment by dstansby** [comment]
Thanks for the PR! Could you put the integer dtype histogram change into a separate PR to make it easier to review?

**Comment by p-j-smith** [comment]
> Could you put the integer dtype histogram change into a separate PR to make it easier to review?

Yep, will do 🙂

File: src/napari_matplotlib/histogram.py [hunk 1]
```diff
@@ -29,26 +39,143 @@ def __init__(
         parent: Optional[QWidget] = None,
     ):
         super().__init__(napari_viewer, parent=parent)
+
+        # Create widgets for setting bin parameters
+        bins_start = QDoubleSpinBox()
+        bins_start.setObjectName("bins start")
+        bins_start.setStepType(QAbstractSpinBox.AdaptiveDecimalStepType)
+        bins_start.setRange(-1e10, 1e10)
```
Comment (p-j-smith) [inline src/napari_matplotlib/histogram.py:47]: I wasn't sure what a reasonable range would be (rather than the default of 0 to 100)?

File: src/napari_matplotlib/histogram.py [hunk 2]
```diff
@@ -29,26 +39,143 @@ def __init__(
...
+        n_decimals = 0 if np.issubdtype(layer_data.dtype, np.integer) else 2
+        is_unsigned = layer_data.dtype.kind == "u"
+        minimum_value = 0 if is_unsigned else -1e10
+
+        bins_start = self.findChild(QDoubleSpinBox, name="bins start")
+        bins_stop = self.findChild(QDoubleSpinBox, name="bins stop")
+        bins_start.setDecimals(n_decimals)
+        bins_stop.setDecimals(n_decimals)
+        bins_start.setMinimum(minimum_value)
+        bins_stop.setMinimum(minimum_value)
```
Comment (p-j-smith) [inline src/napari_matplotlib/histogram.py:162]: it's necessary to set the minimum bin values to 0 for unsigned integers - if negative bin limits are selected for data with unsigned integers, `plt.hist` complains that bins must be monotonically increasing (as there's an integer underflow in the bins produced by`np.linspace`)

Comment (p-j-smith) [reply src/napari_matplotlib/histogram.py:162]: it's not necessary to set the decimals to 0 for integers, but makes it clear to the users that the data are integers and float bin limits are not appropriate

File: src/napari_matplotlib/histogram.py [hunk 3]
```diff
@@ -55,6 +109,31 @@ def on_update_layers(self) -> None:
...
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(True)
+
+        self._bin_widgets["start"].setDecimals(n_decimals)
+        self._bin_widgets["stop"].setDecimals(n_decimals)
+        self._bin_widgets["start"].setMinimum(minimum_value)
+        self._bin_widgets["stop"].setMinimum(minimum_value)
+
+        for widget in self._bin_widgets.values():
+            widget.blockSignals(False)
```
Comment (p-j-smith) [inline src/napari_matplotlib/histogram.py:135]: temporarily disable the callbacks otherwise the plot is re-drawn each time the widget value changes

File: src/napari_matplotlib/histogram.py [hunk 4]
```diff
@@ -63,24 +142,83 @@ def _update_contrast_lims(self) -> None:
...
-        bins = _get_bins(data)
+        if data.dtype.kind in {"i", "u"}:
+            # Make sure integer data types have integer sized bins
+            step = abs(self.bins_stop - self.bins_start) // (self.bins_num)
+            step = max(1, step)
+            bins = np.arange(self.bins_start, self.bins_stop + step, step)
+        else:
+            bins = np.linspace(
+                self.bins_start, self.bins_stop, self.bins_num + 1
+            )
```
Comment (p-j-smith) [inline src/napari_matplotlib/histogram.py:221]: adding +1 to `self.bins_num` as `np.linspace` takes the number of bin edges rather than the number of bins

File: src/napari_matplotlib/histogram.py [hunk 5]
```diff
@@ -63,24 +142,83 @@ def _update_contrast_lims(self) -> None:
...
+        data = self._get_layer_data(layer)
+
         # Important to calculate bins after slicing 3D data, to avoid reading
         # whole cube into memory.
-        bins = _get_bins(data)
+        if data.dtype.kind in {"i", "u"}:
+            # Make sure integer data types have integer sized bins
+            step = abs(self.bins_stop - self.bins_start) // (self.bins_num)
+            step = max(1, step)
+            bins = np.arange(self.bins_start, self.bins_stop + step, step)
```
Comment (p-j-smith) [inline src/napari_matplotlib/histogram.py:217]: I think it would be good to show a notification when the actual number of bins used differs from that requested in the widgets, but perhaps this could be a different PR?

Comment (p-j-smith) [reply src/napari_matplotlib/histogram.py]: I've added the notification now, but happy to remove it and add separately if it's easier

**Comment by dstansby** [comment]
I am finally thinking about this (sorry it took so long!), and my current thoughts are to keep things simple we should just have a setting for the number of bins, and show the full range of data without being able to control the start/stop. I am open to adding confiurable start/stop in the future **if** there's a compelling use case, but for now I think configuring the number of bins is a nice feature that isn't too complicated by itself.

Would you be able to update this PR (or open a new one) to just have the number of bins configurable? Obviously no rush at all 😄, but let me know if it isn't likely to be soon in case I get the bug to implement it

**Comment by p-j-smith** [comment]
Yeah good idea, it became surprisingly complicated! I've removed the notification of requested vs actual bins that I previously added (you can see it in the video above) , do you want it added back?

**Review by dstansby — APPROVED** [review]
Sorry this has taken so long to get to, but I gave it a try and it looks 👍👍👍👍 . I've merged `main` into here, and will update the changelog entry, then get this in a release 😄

**Comment by p-j-smith** [comment]
> Sorry this has taken so long to get to,

no worries 😄  Thanks for the reviews and for merging it!
